### PR TITLE
fix #19279: Memory leak in Zip.cpp

### LINKED ISSUE #19279: Potential memory leak in src/openrct2/core/Zip.cpp
Potential memory leak in [Zip.cpp](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/core/Zip.cpp#L170) line 170. Calling zip_source_buffer() without calling zip_source_free() to free the memory will cause a memory leak. [Doc](https://libzip.org/documentation/zip_source.html) says "[zip_source_free(3)](https://libzip.org/documentation/zip_source_free.html) only needs to be called when these functions ([zip_add, zip_replace](https://libzip.org/documentation/zip_add.html)) return an error."
image

### PR DESCRIPTION
**Pull request by icy17**
Fix #19279 .

### CODE REVIEW

File: src/openrct2/core/Zip.cpp [hunk 1]
```diff
@@ -169,13 +169,18 @@ class ZipArchive final : public IZipArchive
...
+            res = zip_replace(_zip, index.value(), source);
         }
         else
         {
-            zip_add(_zip, path.data(), source);
+            res = zip_add(_zip, path.data(), source);
+        }
+        if (res == -1)
+        {
+            zip_source_free(source);
```
Comment (duncanspumpkin) [inline src/openrct2/core/Zip.cpp:183]: We should also log some sort of error message as well since the add/replace failed.

Comment (IntelOrca) [reply src/openrct2/core/Zip.cpp:183]: Or throw an exception

**Comment by CookiePLMonster** [comment]
`zip_replace` and `zip_add` are deprecated and they both have functionally equivalent non-deprecated alternatives - since it's in the very same function, maybe it's worth changing as part of this PR too?

**Comment by ZehMatt** [comment]
> `zip_replace` and `zip_add` are deprecated and they both have functionally equivalent non-deprecated alternatives - since it's in the very same function, maybe it's worth changing as part of this PR too?

Thats actually a good idea if they are deprecated, we shouldn't be using that then.

**Comment by CookiePLMonster** [comment]
For completeness:

> The function zip_add() is the obsolete version of [zip_file_add(3)](https://libzip.org/documentation/zip_file_add.html). It is the same as calling [zip_file_add(3)](https://libzip.org/documentation/zip_file_add.html) with an empty flags argument. Similarly, the zip_replace() function is the obsolete version of [zip_file_replace(3)](https://libzip.org/documentation/zip_file_replace.html). It is the same as calling [zip_file_replace(3)](https://libzip.org/documentation/zip_file_replace.html) with an empty flags argument.

**Comment by CookiePLMonster** [comment]
@icy17 any update on this?

**Comment by ZehMatt** [comment]
Yeah fair point, lets give it a few more days and if nothing happens I'll rebase and address the issues.

**Comment by tupaschoal** [comment]
@ZehMatt are you going to take over this?

**Comment by ZehMatt** [comment]
> @ZehMatt are you going to take over this?

I'll see if I can get it done this weekend.

File: src/openrct2/core/Zip.cpp [hunk 2]
```diff
@@ -169,13 +169,19 @@ class ZipArchive final : public IZipArchive
...
         }
         else
         {
-            zip_add(_zip, path.data(), source);
+            res = zip_file_add(_zip, path.data(), source, 0);
+        }
+        if (res == -1)
+        {
+            zip_source_free(source);
+            throw std::runtime_error("Unable to set file contents.");
```
Comment (tupaschoal) [inline src/openrct2/core/Zip.cpp:184]: Let's open an issue to improve error reporting here reading from `_zip`

**Pull request merged**